chain/epoch_manager/src/lib.rs

```diff
@@ -1462,73 +1441,150 @@ impl EpochManager {
-        };
-        if &aggregator.epoch_id != epoch_id {
-            aggregator = EpochInfoAggregator::new(epoch_id.clone(), *last_block_hash);
-            epoch_change = true;
+            Ok(self.epoch_info_aggregator.clone())
         }
-        let epoch_info = self.get_epoch_info(epoch_id)?.clone();
-        let mut new_aggregator = EpochInfoAggregator::new(epoch_id.clone(), *last_block_hash);
-        let mut cur_hash = *last_block_hash;
-        let mut overwrite = false;
-        while cur_hash != aggregator.last_block_hash || epoch_change {
-            // Avoid cloning
-            let prev_hash = *self.get_block_info(&cur_hash)?.prev_hash();
-            let prev_height = self.get_block_info(&prev_hash).map(|info| *info.height());
+    }
 
-            let block_info = self.get_block_info(&cur_hash)?;
-            if block_info.epoch_id() != epoch_id || block_info.prev_hash() == &CryptoHash::default()
+    /// Aggregates epoch info between last final block and given block.
+    ///
+    /// More specifically, aggregates epoch information from block denoted by
+    /// `self.epoch_info_aggregator.last_block_hash` (excluding that block) up
+    /// to one denoted by `block_hash` (including that block).  If the two
+    /// blocks belong to different epochs, stops aggregating once it reaches
+    /// start of epoch `block_hash` belongs to.
+    ///
+    /// The block hash passed as argument should be a latest final block or
+    /// a descendant of a latest final block. Calling this method on any other
+    /// block is likely to result in performance issues since handling forks
+    /// will force it to traverse the entire epoch from scratch.
+    ///
+    /// If `block_hash` equals `self.epoch_info_aggregator.last_block_hash`
+    /// returns None.  Otherwise returns `Some((aggregator, full_info))` tuple.
+    /// The first element of the pair is aggregator with collected information;
+    /// the second specifies whether the returned aggregator includes full
+    /// information about an epoch (such that it does not need to be merged with
+    /// `self.epoch_info_aggregator`).  That happens if the method reaches epoch
+    /// boundary.
+    fn aggregate_epoch_info_upto(
+        &mut self,
+        block_hash: &CryptoHash,
+    ) -> Result<Option<(EpochInfoAggregator, bool)>, EpochError> {
+        if block_hash == &self.epoch_info_aggregator.last_block_hash {
+            return Ok(None);
+        }
+
+        if cfg!(debug) {
+            let agg_hash = self.epoch_info_aggregator.last_block_hash.clone();
+            let agg_height = *self.get_block_info(&agg_hash)?.height();
+            let block_height = *self.get_block_info(block_hash)?.height();
+            assert!(
+                agg_height < block_height,
+                "#{} {} >= #{} {}",
+                agg_hash,
+                agg_height,
+                block_hash,
+                block_height
+            );
```

Comment:
```suggestion
            assert!(
                agg_height < block_height,
                "#{agg_hash} {agg_height} >= #{block_hash} {block_height}",
            );
```

doesn't really matter, but we have string interpolation now!

Reply:
Oh, God, I hate that feature.

Reply:
Curious, why? For me, it makes the code so much more readable. Though, the fact that only identifiers are allowed creates a significant discontinuity -- you can't *always* use interpolation. 

It's also implemented in a non-transparent for tooling way, which makes IDE life harder

Reply:
It magically pulls in variables from the local scope with no indication that it’s what it’s doing.  Why is `let bar = format!("{foo}");` interpolating variable `foo` but `let bar = "{foo}";` is just a string?  How do I know if `some_random_macro!("{foo}");` will interpolate the variable?

Reply:
Heh yes, that's exactly what IDE struggle with -- it'd be much better if it were first-class syntax, like in Kotlin.

Though, I must say the same is true of the usual interpolation: `"{}"` vs `format!("{}")`.

Syntax highlighting works-around here a bunch (well, with ligatures me):

![image](https://user-images.githubusercontent.com/1711539/162425393-6b42fb79-d729-457e-964f-b572c54eb6e8.png)

That should be available in Emacs as well: https://github.com/emacs-lsp/lsp-mode/pull/1721

But yeah, it's a shame that we need to expand macros to figure that out.